### PR TITLE
added serialize run functionality

### PR DESCRIPTION
**Pull request by janvanrijn**
#### What does this PR implement/fix? Explain your changes.
It allows run objects (including predictions and traces) to be serialized to disk, and reloaded. This functionality is almost used in all my projects, and I can imagine us using this for the benchmark study. 

#### How should this PR be tested?
Unit tests should pass, code should make sense, please check the unit test if run equality check is OK.

### CODE REVIEW

File: openml/runs/run.py [hunk 1]
```diff
@@ -65,6 +66,49 @@ def __str__(self):
     def _repr_pretty_(self, pp, cycle):
         pp.text(str(self))
 
+    @classmethod
+    def from_filesystem(cls, folder):
```
Comment (mfeurer) [inline openml/runs/run.py]: Could you please add a docstring?

File: openml/runs/run.py [hunk 2]
```diff
@@ -65,6 +66,49 @@ def __str__(self):
...
+
+        if os.path.isfile(trace_path):
+            with open(trace_path, 'r') as fp:
+                trace = arff.load(fp)
+                run.trace_attributes = trace['attributes']
+                run.trace_content = trace['data']
+
+        return run
+
+    def to_filesystem(self, output_directory):
```
Comment (mfeurer) [inline openml/runs/run.py]: Could you please add a docstring, here, too?

File: openml/runs/run.py [hunk 3]
```diff
@@ -65,6 +66,49 @@ def __str__(self):
...
+        run_xml = self._create_description_xml()
+        predictions_arff = arff.dumps(self._generate_arff_dict())
+
+        with open(output_directory + '/description.xml', 'w') as f:
```
Comment (mfeurer) [inline openml/runs/run.py]: Could you please use `os.path.join` as above?

File: openml/runs/run.py [hunk 4]
```diff
@@ -65,6 +66,49 @@ def __str__(self):
...
+            f.write(run_xml)
+        with open(output_directory + '/predictions.arff', 'w') as f:
```
Comment (mfeurer) [inline openml/runs/run.py]: Same here.

File: openml/runs/run.py [hunk 5]
```diff
@@ -65,6 +66,49 @@ def __str__(self):
...
+            f.write(predictions_arff)
+
+        if self.trace_content is not None:
+            trace_arff = arff.dumps(self._generate_trace_arff_dict())
+            with open(output_directory + '/trace.arff', 'w') as f:
```
Comment (mfeurer) [inline openml/runs/run.py]: And here

File: tests/test_runs/test_run.py [hunk 6]
```diff
@@ -66,3 +70,55 @@ def test_tagging(self):
...
+
+        task = openml.tasks.get_task(119)
+        run = openml.runs.run_model_on_task(task, model)
+
+        cache_path = os.path.join(self.workdir, 'runs', str(random.getrandbits(128)))
+        os.makedirs(cache_path)
+        run.to_filesystem(cache_path)
+
+        run_prime = openml.runs.OpenMLRun.from_filesystem(cache_path)
+        self._test_run_obj_equals(run, run_prime)
```
Comment (mfeurer) [inline tests/test_runs/test_run.py]: You should add a check here that the trace is available. The function `_test_run_obj_equals` does not guarantee this.

**Comment by janvanrijn** [comment]
Agreed to all.

File: openml/runs/run.py [hunk 7]
```diff
@@ -68,6 +69,21 @@ def _repr_pretty_(self, pp, cycle):
 
     @classmethod
     def from_filesystem(cls, folder):
+        """
+        The inverse of the to_filesystem method. Initiates a run based
```
Comment (mfeurer) [inline openml/runs/run.py]: I think that `initiate` is unfortunate wording here as it also means 'to start'. How about 'instantiates a run object'?

**Comment by mfeurer** [comment]
I did not yet merge this to develop/master as you requested in my other PR because we can do this whenever we're ready to start up some experiments. I you need this badly we can merge this into master under a `0.7.1` version tag.

**Comment by janvanrijn** [comment]
Sure. Where can I find the change log?

**Comment by janvanrijn** [comment]
Ouch, I extended the unit tests with a publish statement, and apparently the model is always needed to be present in order to be able to upload. Don't know if we should really enforce this, as it seems just a sanity check, but I encountered some more discrepancies. Apparently the to/from xml functions upon which this relied did not work perfectly. 

I pushed a fix, also adding more checks to both the serialize/unserialize functions and unit tests. However, I have a feeling the Run to/from XML functionality could use some more extensive unit tests. Please have a critical look at this and feel free to extend.

**Comment by mfeurer** [comment]
The changelog is a bit hidden and we need to re-launch it, but it's here: https://github.com/openml/openml-python/blob/master/doc/progress.rst.

Regarding the behavior of publish when no flow is present, I'm not sure if/how we can support this. Instead of having a flow in the run, we would keep the model in the run and change this once we want to upload the run to OpenML? @joaquinvanschoren also opened an issue about this: https://github.com/openml/openml-python/issues/457

**Comment by janvanrijn** [comment]
This is different. The publish function obviously requires a flow id (which is why Joaquin opened the issue), but requires a model (which I didn't store). Not sure if we really should require this, but this is what I fixed in the last PR. 

Good to merge?

**Comment by janvanrijn** [comment]
Almost forgot, but it's now in the change log 

Made an issue with some suggestions #465

**Comment by mfeurer** [comment]
I think you forgot to push the commit.

**Comment by janvanrijn** [comment]
I accidentally put it in the other pull request (listing)

**Pull request merged**